File: go/collatz-conjecture/collatz_conjecture.go
package collatzconjecture

import "errors"

func CollatzConjecture(n int) (int, error) {
	if n <= 0 {
		return 0, errors.New("Input needs to be positive")
	}

	return runStep(n, 0), nil
}

func runStep(n int, steps int) int {
	if n == 1 {
		return steps
	}

    var newVal int
    if n%2 == 0 {
        newVal = n / 2
	} else {
        newVal = n * 3 + 1
	}

    steps++

    return runStep(newVal, steps)
}

File: elixir/collatz-conjecture/lib/collatz_conjecture.ex
defmodule CollatzConjecture do
  @doc """
  calc/1 takes an integer and returns the number of steps required to get the
  number to 1 when following the rules:
    - if number is odd, multiply with 3 and add 1
    - if number is even, divide by 2
  """
  @spec calc(input :: pos_integer()) :: non_neg_integer()
  def calc(1), do: 0
  def calc(input), do: calc(input, 0)
  def calc(1, step_count), do: step_count
  def calc(input, step_count) do
    input
    |> run_step()
    |> calc(step_count + 1)
  end

  defp run_step(num) when is_integer(num) and num > 0 and rem(num, 2) == 0, do: div(num, 2)
  defp run_step(num) when is_integer(num) and num > 0, do: num * 3 + 1
end
